libs/execution_context/ebpf_program.c

```diff
@@ -447,79 +451,38 @@ ebpf_program_create(_Outptr_ ebpf_program_t** program)
     ebpf_list_initialize(&local_program->links);
     ebpf_lock_create(&local_program->lock);
 
-    retval = ebpf_object_initialize(
-        &local_program->object, EBPF_OBJECT_PROGRAM, _ebpf_program_free, _ebpf_program_get_program_type);
-    if (retval != EBPF_SUCCESS) {
-        goto Done;
-    }
-
     local_program->bpf_prog_type = BPF_PROG_TYPE_UNSPEC;
 
-    *program = local_program;
-    local_program = NULL;
-    retval = EBPF_SUCCESS;
-
-Done:
-    if (local_program) {
-        _ebpf_program_epoch_free(local_program);
-    }
-
-    EBPF_RETURN_RESULT(retval);
-}
-
-_Must_inspect_result_ ebpf_result_t
-ebpf_program_initialize(_Inout_ ebpf_program_t* program, _In_ const ebpf_program_parameters_t* program_parameters)
-{
-    EBPF_LOG_ENTRY();
-    ebpf_result_t return_value;
-    ebpf_utf8_string_t local_program_name = {NULL, 0};
-    ebpf_utf8_string_t local_section_name = {NULL, 0};
-    ebpf_utf8_string_t local_file_name = {NULL, 0};
-    uint8_t* local_program_info_hash = NULL;
-
-    ebpf_lock_state_t state = ebpf_lock_lock(&program->lock);
-    bool lock_held = true;
-
-    if (program->parameters.code_type != EBPF_CODE_NONE) {
-        EBPF_LOG_MESSAGE_UINT64(
-            EBPF_TRACELOG_LEVEL_ERROR,
-            EBPF_TRACELOG_KEYWORD_PROGRAM,
-            "ebpf_program_initialize program->parameters.code_type must be EBPF_CODE_NONE",
-            program->parameters.code_type);
-        return_value = EBPF_INVALID_ARGUMENT;
-        goto Done;
-    }
-
     if (program_parameters->program_name.length >= BPF_OBJ_NAME_LEN) {
         EBPF_LOG_MESSAGE_UINT64(
             EBPF_TRACELOG_LEVEL_ERROR,
             EBPF_TRACELOG_KEYWORD_PROGRAM,
             "Program name must be less than BPF_OBJ_NAME_LEN",
             program_parameters->program_name.length);
-        return_value = EBPF_INVALID_ARGUMENT;
+        retval = EBPF_INVALID_ARGUMENT;
         goto Done;
     }
 
-    return_value = ebpf_duplicate_utf8_string(&local_program_name, &program_parameters->program_name);
-    if (return_value != EBPF_SUCCESS) {
+    retval = ebpf_duplicate_utf8_string(&local_program_name, &program_parameters->program_name);
```

Comment:
Do you need to change the SAL annotation that program_parameters need not be protected by the lock?

Reply:
That's why I filed #2325. The exact rules about when it's safe to access fields in ebpf_program_t is unclear. Some fields are protected by ebpf_program_t::lock and others are protected by the rundown protection on the program information provider. Until we document and audit the carefully, we can't be sure it's race free.

Reply:
But won't SAL complain about the changes you made in this PR?

Reply:
Doesn't appear to.